include/kvs_async_client.hpp

```diff
@@ -0,0 +1,548 @@
+    // handle a key address response
+    if (pollitems_[0].revents & ZMQ_POLLIN) {
+      string serialized = kZmqUtil->recv_string(&key_address_puller_);
+      KeyAddressResponse response;
+      response.ParseFromString(serialized);
+      Key key = response.addresses(0).key();
+      if (pending_request_map_.find(key) != pending_request_map_.end()) {
+        if (response.error() == 1) {
+          log_->error(
+              "No servers have joined the cluster yet. Retrying request.");
+          pending_request_map_[key].first = std::chrono::system_clock::now();
+          query_routing_async(key);
+        } else {
+          // populate cache
+          for (const Address& ip : response.addresses(0).ips()) {
+            key_address_cache_[key].insert(ip);
+          }
+          // handle stuff in pending request map
+          for (auto& pair : pending_request_map_[key].second) {
+            try_request(pair.first, pair.second);
+          }
+          // GC the pending request map
+          pending_request_map_.erase(key);
+        }
+      }
+    }
+    // handle a response
+    if (pollitems_[1].revents & ZMQ_POLLIN) {
+      string serialized = kZmqUtil->recv_string(&response_puller_);
+      KeyResponse response;
+      response.ParseFromString(serialized);
+      Key key = response.tuples(0).key();
+      if (response.type() == RequestType::GET) {
+        // GET
+        if (pending_get_response_map_.find(key) !=
+            pending_get_response_map_.end()) {
+          if (check_tuple(response.tuples(0))) {
+            // error no == 2, so re-issue request
+            pending_get_response_map_[key].tp_ =
+                std::chrono::system_clock::now();
+            try_request(
+                pending_get_response_map_[key].request_addr_pair_.first,
+                pending_get_response_map_[key].request_addr_pair_.second);
+          } else {
+            // error no == 0 or 1
+            result.push_back(
+                {response,
+                 pending_get_response_map_[key].request_addr_pair_.second});
+            // GC
+            pending_get_response_map_.erase(key);
+          }
+        }
+      } else {
+        // PUT
+        if (pending_put_response_map_.find(key) !=
+                pending_put_response_map_.end() &&
+            pending_put_response_map_[key].find(response.response_id()) !=
+                pending_put_response_map_[key].end()) {
+          if (check_tuple(response.tuples(0))) {
+            // error no == 2, so re-issue request
+            pending_put_response_map_[key][response.response_id()].tp_ =
+                std::chrono::system_clock::now();
+            try_request(pending_put_response_map_[key][response.response_id()]
+                            .request_addr_pair_.first,
+                        pending_put_response_map_[key][response.response_id()]
+                            .request_addr_pair_.second);
+          } else {
+            // error no == 0 or 1
```

Comment:
The KVS should never return `errno = 1` for a PUT request, no?

Reply:
right

Reply:
Okay, just making sure. Let's clean up the comments for clarity.